### PR TITLE
Wip master

### PR DESCRIPTION
**Pull request by jordanhudgens**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,58 +1,148 @@
...
-7. Explain the problems with form_tag for an edit page
+Now that you know how to implement a create action the next logical step will be to integrate an edit/update action. As you may have noticed, there is a trend in Rails conventions where the logic for rendering a form is separate from the action that the manages the database record alteration. For example:
 
+* The `new` action in the controller simply rendering the `new` form
 
-## Notes
+* The `create` action is what actually handled the process for inserting the form data into the database
 
-let's build an edit and update form that uses form_tag
+In like fashion, the `edit` and `update` actions have a similar convention:
```
Comment (jmburges) [inline README.md:11]: Take what they know, and bring them to what they will learn! :+1:

File: README.md [hunk 2]
```diff
@@ -1,58 +1,148 @@
...
-first, what's the URL or named route for the form_tag? post_path(@post). submit the form with that, we get to post show because the form submitted with a GET/POST request instead of a PUT request.
+To start off, let's draw a `get` route for our edit form, since the form will need to know which record is being edited this will need to be a dynamic route that accepts an `:id` as a parameter so it can be accessed by the controller:
 
-the _input hack passing :method to form_tag automatically embeds input hack in your form.
+```ruby
+get '/post/:id/edit', to: 'posts#edit'
+```
 
-now that we have the form wrapper pointing to the write action/url with the method hack, what else is different about our edit form vs our new form?
+We still need to draw one additional route to handle the `update` action, this route will also need to be dynamic and accept an `:id` as a parameter so the `update` action will know what record is being altered:
```
Comment (jmburges) [inline README.md]: also mention that since this is sending data to the server it's not GET, since it's not creating a new record it's not POST, so it's PUT

File: README.md [hunk 3]
```diff
@@ -1,58 +1,148 @@
...
+def edit
+end
 
-What does update need to do in the action?
+def update
+end
+```
 
-similar to edit, the first step is to simply find the post we're talking about.
+And then create the edit view template `touch app/views/posts/edit.html.erb`
```
Comment (jmburges) [inline README.md]: Just say create a file called blah

File: README.md [hunk 4]
```diff
@@ -1,58 +1,148 @@
...
 
-once we found it, how can we update that instance?
+In the future we will call the form partial to integrate the form, but for this lesson let's just copy and paste the form code from the `new` form:
```
Comment (jmburges) [inline README.md]: Don't mention partials yet

File: README.md [hunk 5]
```diff
@@ -1,58 +1,148 @@
...
+If you open the browser and go to the `edit` page it will now display the form, but you may have noticed a pretty big flaw, it doesn't load the record's data into the form, there are a few tasks that we'll need to do in order to implement this behavior, first let's have our `edit` action store the `post` record in an instance variable:
 
-in the event you only edit the title, does the post content get updated if we are always assigning all fields? dirty_attributes like content_changed? internalized in the sql for save/update (is that true?)
+```ruby
+def edit
+  @post = Post.find(params[:id])
+end
+```
+
+Now that the `edit` view template will have access to the `@post` object, we need to refactor the form so that it auto fills the form fields with the data from the `post` record, this is done below:
```
Comment (jmburges) [inline README.md:80]: Just re-show the code updated, it's a bit hard to follow

File: README.md [hunk 6]
```diff
@@ -1,58 +1,148 @@
...
+<% # app/views/posts/edit.html.erb %>
+
+...
+<%= text_field_tag :title, @post.title %><br>
+...
+<%= text_area_tag :description, @post.description %><br>
+...
+```
+
+This will now populate the form, but if you tried to submit the form you may notice that it's redirecting to the show page and not changing the values. We need to first make a change to the form so that it knows what route the data should be passed through as well as what HTTP verb needs to be called, make the following change to the `form_tag` line:
```
Comment (jmburges) [inline README.md]: Point out that the path for update is the same post_path, just the method is different. Which makes sense because we are working with the same resource, just doing a different action.

Ugh I love the HTTP protocol. Tim Berners-Lee is such a genius

File: README.md [hunk 7]
```diff
@@ -1,58 +1,148 @@
...
+
+```ERB
+<%= form_tag post_path(@post), method: "put" do %>
+```
+
+By explicitly stating the route and method the form will now be routed to the `update` action. Before we try to implement the `update` action, let's first make sure the data is being routed properly. Enter the following code inside of the `update` method:
```
Comment (jmburges) [inline README.md:104]: Explain briefly what `raise` does.

Comment (jmburges) [inline README.md:104]: Also another option is to just `puts` it and show it shows up in the rails c

File: README.md [hunk 8]
```diff
@@ -1,58 +1,148 @@
...
+end
+```
+
+If you open up the browser and navigate to an edit page, such as: `localhost:3000/post/5/edit` and change some elements in and form and submit it, it should take you to an error page that prints out the params from the form, such as the below image:
+
+![Raised Exception for Update Action](http://reif.io/lib/flatiron/update_raised_exception.png)
+
+As you can see, the parameters are being passed to the update action, with that in mind let's implement the functionality needed inside of the `update` action so that it will take the form data and update the specified record. Let's write some pseudo code for what the `update` action should do:
+
+* Query the database for the `Post` record that matched the `:id` passed to the route
```
Comment (jmburges) [inline README.md:120]: Love the pseudocode

File: README.md [hunk 9]
```diff
@@ -1,58 +1,148 @@
...
+* Store the query in an instance variable
+
+* Update the values passed from the form
+
+* Save the changes in the database
+
+* Redirect the user to the `show` page so they can see the updated record
+
+
+We'll take advantage of the `update` method so that we're not manually assigning each attribute:
```
Comment (jmburges) [inline README.md:131]: clarify you mean the `update` method provided by ActiveRecord. Don't want people to confuse the `update` you're writing and the AR update

Comment (jmburges) [inline README.md:131]: Also quick explanation on how the `update` method works. (pretty much, it takes in a hash formatted like this (hash) OH HEY look that's the same as params)

**Comment by jmburges** [comment]
back at you

**Comment by jmburges** [comment]
Disable Strong Params!

**Comment by jordanhudgens** [comment]
@jmburges This has been updated.

**Comment by jmburges** [comment]
Make sure you are only using `form_tag` and that you include the rails project in here with strong params disabled!

**Comment by jmburges** [comment]
then merge away
